exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java

```diff
@@ -408,6 +424,9 @@ private void runPhysicalPlan(final PhysicalPlan plan) throws ExecutionSetupExcep
     fragmentsRunner.setFragmentsInfo(work.getFragments(), work.getRootFragment(), work.getRootOperator());
 
     startQueryProcessing();
+    if (enableRuntimeFilter) {
+      runtimeFilterManager.waitForComplete();
```

Comment:
Why is this needed here ?

Reply:
To make sure the `RuntimerFilterManager`'s sending out  `ByteBuf`s to be safely cleared out no matter the network successed or failed.

Reply:
But based on my understanding `RuntimeFilterManager` only sends out something over network during broadcasting of aggregated bloomfilter. Broadcast currently happens on the Rpc thread which needs to be changed. So at what time will this Foreman thread will end up also sending the buffers over network ?

Reply:
There is various cases that the network between the foreman and receivers break down. Also the norm case is the receiver finished earlier before receiving the bloom filter. So the security way is to use the `SendingAccountor` to release the allocated `ByteBuf`.

Reply:
`SendingAccountor` on which this Foreman thread waits for completion is local to the `RunTimeFilterManager`. The job of `SendingAccountor` is to wait the caller thread iff there is any pending message to be sent unless the send is completed successfully or failed.  The buffer release is handled in the failure path of the mesaage not in sending accountor itself. The only time this `SendingAccountor` is incremented is when there is aggregated filter to be broadcasted. 
By the time this `Foreman` thread is exited there is no guarantee that remote minor fragments may have completed sending all their respective `Bloomfilters` and that `aggregated filter` is also broadcasted, hence it will be no-op in those cases. And more over if there is any failure case as you mentioned above there are chances of memory leak. To avoid that it will make more sense to put this `runtimeFilterManager.waitForComplete` inside `ForemanResult::close` method right before `closeFuture.removeListener(closeListener)`

Reply:
sorry for the wrong state. will change it to the right position .